server/e2e_test_from_label.go

```diff
@@ -0,0 +1,102 @@
+package server
+
+import (
+	"context"
+
+	"github.com/google/go-github/v39/github"
+	"github.com/mattermost/mattermost-mattermod/model"
+	"github.com/mattermost/mattermost-server/v5/mlog"
+)
+
+const (
+	e2eTestFromLabelMsgPRNotMergeable   = "E2E tests not automatically triggered, because the PR is not in a mergeable state. Please update the branch with the base branch and resolve outstanding conflicts."
+	e2eTestFromLabelMsgPRHasNoApprovals = "E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label. "
+)
+
+func (e *E2ETestFromLabelError) Error() string {
+	switch e.source {
+	case e2eTestFromLabelMsgPRNotMergeable:
+		return "e2e not running, pr not in mergeable state"
+	case e2eTestFromLabelMsgPRHasNoApprovals:
+		return "e2e not running, pr has no approval"
+	default:
+		panic("unhandled error type")
+	}
+}
+
+type E2ETestFromLabelError struct {
+	source string
+}
+
+func (s *Server) triggerE2ETestFromLabel(ctx context.Context, pr *model.PullRequest) {
+	var e2eTestFromLabelErr *E2ETestFromLabelError
+	defer func() {
+		if e2eTestFromLabelErr != nil {
+			if err := s.sendGitHubComment(ctx, pr.RepoOwner, pr.RepoName, pr.Number, e2eTestFromLabelErr.source); err != nil {
+				mlog.Warn("Error while commenting", mlog.Err(err))
+			}
+		}
+	}()
+
+	prReviews, _, err := s.GithubClient.PullRequests.ListReviews(ctx, pr.RepoOwner, pr.RepoName, pr.Number, nil)
+	if err != nil {
+		mlog.Error("Error getting reviews for the PR",
+			mlog.Int("pr", pr.Number),
+			mlog.String("repo", pr.RepoName),
+			mlog.Err(err))
+		return
+	}
+	if !hasAtLeastOneApproval(prReviews) {
+		e2eTestFromLabelErr = &E2ETestFromLabelError{source: e2eTestFromLabelMsgPRHasNoApprovals}
+		mlog.Warn("Not triggering E2E test, due to missing required approvals",
+			mlog.Int("pr", pr.Number),
+			mlog.String("repo", pr.RepoName),
+			mlog.Err(e2eTestFromLabelErr))
+		return
```

Comment:
@mvitale1989 @metanerd How are we going to deal with error which caused not to run E2E automatically so that a PR will not miss E2E run before merge?

Reply:
We are not dealing with that at the moment. Let's deal later with it, when we see it happening? I have an idea, but we have to be able to fully rely on our infrastructure for the tests, otherwise teams will complain to not being able to merge.